reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,84 +83,124 @@ object StreamSubscriber {
+          case Uninitialized =>
+            Idle(s) -> F.unit
+          case Idle(prev) =>
+            Idle(prev) -> F.delay(s.cancel)
+          case Receiving(prev) =>
+            Receiving(prev) -> F.delay(s.cancel)
           case o =>
             val err = new Error(s"received subscription in invalid state [$o]")
             o -> (F.delay(s.cancel) >> F.raiseError(err))
         }
         case OnNext(a) => {
-          case WaitingOnUpstream(s, r) => Idle(s) -> r.complete(a.some.asRight)
-          case DownstreamCancellation  => DownstreamCancellation -> F.unit
-          case o                       => o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
+          case Receiving(s) =>
+            Receiving(s) -> q.enqueue1(a.some.asRight)
+          case Idle(s) =>
+            Receiving(s) -> q.enqueue1(a.some.asRight)
+          case DownstreamCancellation =>
+            DownstreamCancellation -> F.unit
+          case o =>
+            o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
         }
         case OnComplete => {
-          case WaitingOnUpstream(_, r) => UpstreamCompletion -> r.complete(None.asRight)
-          case _                       => UpstreamCompletion -> F.unit
+          case Receiving(_) =>
+            UpstreamCompletion -> q.enqueue1(None.asRight)
+          case Idle(_) =>
+            UpstreamCompletion -> q.enqueue1(None.asRight)
+          case _ =>
+            UpstreamCompletion -> F.unit
         }
         case OnError(e) => {
-          case WaitingOnUpstream(_, r) => UpstreamError(e) -> r.complete(e.asLeft)
-          case _                       => UpstreamError(e) -> F.unit
+          case Receiving(_) =>
+            UpstreamError(e) -> (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight))
+          case _ =>
+           UpstreamError(e) -> F.unit
         }
         case OnFinalize => {
-          case WaitingOnUpstream(sub, r) =>
-            DownstreamCancellation -> (F.delay(sub.cancel) >> r.complete(None.asRight))
-          case Idle(sub) => DownstreamCancellation -> F.delay(sub.cancel)
-          case o         => o -> F.unit
+          case Idle(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
+          case Receiving(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
+          case o =>
+            o -> F.unit
         }
-        case OnDequeue(r) => {
-          case Uninitialized          => RequestBeforeSubscription(r) -> F.unit
-          case Idle(sub)              => WaitingOnUpstream(sub, r) -> F.delay(sub.request(1))
-          case err @ UpstreamError(e) => err -> r.complete(e.asLeft)
-          case UpstreamCompletion     => UpstreamCompletion -> r.complete(None.asRight)
-          case o                      => o -> r.complete((new Error(s"received request in invalid state [$o]")).asLeft)
+        case OnDequeue => {
```

Comment:
I think the only states we need to handle here are `Uninitialized` and `Idle`, everything else is a noop.

We already handle errors and terminating the queue when processing `OnComplete` and `OnError` cases, so it doesn't seem like there is a need to repeat that here. Have I overlooked something?